src/test/java/BaseTest.java

```diff
@@ -1,11 +1,260 @@
+        playListFromSideBar.click();
+    }
+
+    protected void doubleClickAPlaylist() {
+        action = new Actions(driver);
+        WebElement playListFromSideBar = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector(".playlist:nth-child(5)")));
+        action.doubleClick(playListFromSideBar);
+    }
+
+
+    protected void deleteSelectedPlayList() {
+        WebElement deleteButton = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector(".btn-delete-playlist")));
+        deleteButton.click();
+    }
+
+    protected void managePopUpWindow() {
+        //WebElement confirm = driver.findElement(By.cssSelector("button.ok"));
+        //f (Assert.assertTrue(driver.findElement(By.cssSelector("button.ok")).isDisplayed()) {
+        //driver.findElement(By.cssSelector("button.ok")).click();
+        // confirm.click();
+        //}
+        try {
+            WebElement confirm = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector("button.ok")));
+            confirm.click();
+        } catch(Exception e) {
+
+        }
+    }
+    protected String getSongDeletedFromPlaylistNotification() {
+        WebElement deletedFromPlaylistNotification = wait.until(ExpectedConditions.elementToBeClickable(By.cssSelector("div.success.show")));
+        return deletedFromPlaylistNotification.getText();
+    }
+
+    //count songs helper methods
+    protected void choosePlaylistByName(String playlistName) {
+    wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//a[contains(text(),'" + playlistName + "')]"))).click();
+    }
+
+    public int countSongs() {
+    return driver.findElements(By.cssSelector("section#playlistWrapper td.title")).size() ;
+    }
+
+    protected String getPlaylistDetails() {
+    return wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("span.meta.text-secondary span.meta"))).getText();
+    }
+
+    protected void displayAllSongs() {
+        List<WebElement> songList = driver.findElements(By.cssSelector("section#playlistWrapper td.title"));
+        System.out.println("Number of Songs Found:" + countSongs());
+        for (WebElement e: songList) {
+            System.out.println(e.getText());
+        }
+    }
+
+    //rename playlist helper methods
+    String newPlayListName = "re-named";
+    protected void enterNewPlayListName(String newPlayListName) {
+        WebElement playlistInputField = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("[name='name']]")));
+        playlistInputField.sendKeys(Keys.chord(Keys.CONTROL, "A", Keys.BACK_SPACE));
+        playlistInputField.sendKeys(newPlayListName);
+        playlistInputField.sendKeys(Keys.ENTER);
+    }
+
+    public boolean doesPlayListExist() {
+        WebElement playListElement = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//a[text()='" + newPlayListName + "']")));
+        return playListElement.isDisplayed();
+    }
```

Comment:
1. **BasePage and BaseTest** 

**BaseTest** — Implements the necessary methods for work with webdriver.

The BaseTest class contains common setup and teardown methods, as well as any shared functionality or configurations required by multiple test classes.

**BaseTest**

The problem we are trying to solve here is the reuse of common behaviors across test classes avoiding coding duplication and centralizing these actions at one point.

**@BeforeSuite**	- The annotated method will be run before all tests in this suite have run.

**@BeforeMethod** -	The annotated method will be run before all the test methods in the current class have been run.

**@AfterMethod**	- The annotated method will be run after each test method.

_Example without a Base Test Class_

![image](https://user-images.githubusercontent.com/104761482/233972627-2d7ffdc7-6a2f-4e9f-8a3a-a1b4a7a47bc4.png)


**BasePage**  — class serves as a foundation or parent class for individual page classes. It contains common methods and functionalities that are shared across multiple pages in an application.

_Example without a Base Page Class_

![image](https://user-images.githubusercontent.com/104761482/236891493-7f7013c7-fde2-4bf9-9084-3aa7483df8e3.png)

Page Object Model (POM): recommended to associate the methods with the corresponding page class where they are relevant. In the POM, each page or section of a web application is represented by a separate class. These classes encapsulate the elements and actions specific to that page or section.

Move these methods to the corresponding class according to the POM.